pkg/service/v0/service.go

```diff
@@ -0,0 +1,205 @@
+package svc
+
+import (
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"net/http"
+	"path/filepath"
+	"time"
+
+	"github.com/dgrijalva/jwt-go"
+	"github.com/go-chi/chi"
+	"github.com/go-chi/chi/middleware"
+	"github.com/owncloud/ocis-wopiserver/pkg/assets"
+	"github.com/owncloud/ocis-wopiserver/pkg/config"
+	"github.com/owncloud/ocis/ocis-pkg/log"
+	ocsm "github.com/owncloud/ocis/ocis-pkg/middleware"
+)
+
+// Service defines the extension handlers.
+type Service interface {
+	ServeHTTP(http.ResponseWriter, *http.Request)
+}
+
+// NewService returns a service implementation for Service.
+func NewService(opts ...Option) Service {
+	options := newOptions(opts...)
+
+	m := chi.NewMux()
+	m.Use(options.Middleware...)
+
+	m.Use(ocsm.Static(
+		options.Config.HTTP.Root,
+		assets.New(
+			assets.Logger(options.Logger),
+			assets.Config(options.Config),
+		),
+		options.Config.HTTP.CacheTTL,
+	))
+
+	svc := WopiServer{
+		logger: options.Logger,
+		config: options.Config,
+		mux:    m,
+	}
+
+	m.Route(options.Config.HTTP.Root, func(r chi.Router) {
+		r.NotFound(svc.NotFound)
+		r.Use(middleware.StripSlashes)
+		r.Get("/api/v0/wopi/open", svc.OpenFile)
+
+	})
+
+	return svc
+}
+
+// WopiServer defines implements the business logic for Service.
+type WopiServer struct {
+	logger log.Logger
+	config *config.Config
+	mux    *chi.Mux
+}
+
+// ServeHTTP implements the Service interface.
+func (p WopiServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	p.mux.ServeHTTP(w, r)
+}
+
+func (p WopiServer) NotFound(w http.ResponseWriter, r *http.Request) {
+	w.WriteHeader(http.StatusNotFound)
+}
+
+type WopiResponse struct {
+	WopiAccessToken string `json:"accesstoken"`
+	WopiClientURL   string `json:"wopiclienturl"`
+}
+
+type WopiToken struct {
+	ViewMode  string `json:"viewmode"`
+	Token     string `json:"userid"`
+	StorageID string `json:"endpoint"`
+	FolderURL string `json:"folderurl"`
+	FilePath  string `json:"filename"`
+	UserName  string `json:"username"`
+	jwt.StandardClaims
+}
+
+func (p WopiServer) OpenFile(w http.ResponseWriter, r *http.Request) {
+	filePath := r.URL.Query().Get("filePath")
+	if filePath == "" {
+		w.WriteHeader(http.StatusBadRequest)
+		p.logger.Err(errors.New("filePath parameter missing in request"))
+		return
+	}
+
+	token := r.Header.Get("X-Access-Token")
+	if token == "" {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	mode := "write" // TODO: decide how to open it
+
+	viewmode := ""
+
+	switch mode {
+	case "view":
+		viewmode = "VIEW_MODE_VIEW_ONLY"
+	case "read":
+		viewmode = "VIEW_MODE_READ_ONLY"
+	case "write":
+		viewmode = "VIEW_MODE_READ_WRITE"
+	case "default":
+		return
+	}
+
+	wt := WopiToken{
+		ViewMode:  viewmode,
+		Token:     token,
+		StorageID: "1284d238-aa92-42ce-bdc4-0b0000009157", // TODO: do not hardcode
+		FolderURL: "",
+		FilePath:  filePath,
+		UserName:  "Einstein", // TODO: get user from context
```

Comment:
We need a follow-up issue to tackle this and the hardcoded StorageID, or it will probably break silently

Reply:
Was added to https://github.com/owncloud/ocis/issues/2005 from where follow up tickets will be created